dll/win32/lpk/lpk.c

```diff
@@ -132,28 +143,38 @@ LpkGetCharacterPlacement(
     lpResults->nGlyphs = (UINT)cGlyphs;
 
     if (lpResults->lpGlyphs)
-        wcscpy(lpResults->lpGlyphs, lpGlyphs);
-
-    if (lpResults->lpDx && !(dwFlags & GCP_GLYPHSHAPE))
     {
-        int c;
-        for (i = 0; i < nSet; i++)
-        {
-            if (GetCharWidth32W(hdc, lpResults->lpOutString[i], lpResults->lpOutString[i], &c))
-                lpResults->lpDx[i] = c;
-        }
+        if (lpGlyphs)
+            wcscpy(lpResults->lpGlyphs, lpGlyphs);
+
+        else if (lpResults->lpOutString)
+            GetGlyphIndicesW(hdc, lpResults->lpOutString, nSet, lpResults->lpGlyphs, 0);
     }
 
-    /* If glyph shaping was requested */
-    else if (lpResults->lpDx && (dwFlags & GCP_GLYPHSHAPE))
+    if (lpResults->lpDx)
     {
-        int c;
+        /* If glyph shaping was requested */
+        if(dwFlags & GCP_GLYPHSHAPE)
+        {
+            int c;
+
+            if (lpResults->lpGlyphs)
+            {
+                for (i = 0; i < lpResults->nGlyphs; i++)
+                {
+                    if (GetCharWidth32W(hdc, lpResults->lpGlyphs[i], lpResults->lpGlyphs[i], &c))
+                        lpResults->lpDx[i] = c;
```

Comment:
What is `lpDx` initialized to? (e.g. what will be in `lpDx[i]` if `GetCharWidth32W` fails?